pennylane/workflow/execution.py

```diff
@@ -203,15 +212,15 @@ def cost_fn(params, x):
     config = _resolve_execution_config(config, device, tapes, transform_program=transform_program)
 
     transform_program = transform_program or qml.transforms.core.TransformProgram()
-    transform_program, inner_transform = _setup_transform_program(
+    outer_transform, inner_transform = _setup_transform_program(
         transform_program, device, config, cache, cachesize
     )
 
     #### Executing the configured setup #####
-    tapes, post_processing = transform_program(tapes)
+    tapes, outer_post_processing = outer_transform(tapes)
 
-    if transform_program.is_informative:
-        return post_processing(tapes)
+    if outer_transform.is_informative:
+        return user_post_processing(outer_post_processing(tapes))

```

Comment:
Not sure we really need this block anymore. The `outer_transform` will really only ever contain the device preprocessing, which will never be informative.  So I'd say we can delete this block.
```suggestion
    assert not outer_transform.is_informative, "should only contain device preprocessing"
```